src/util/simplify_expr_pointer.cpp

```diff
@@ -420,37 +421,64 @@ bool simplify_exprt::simplify_pointer_offset(exprt &expr)
 static exprt simplify_inequality_address_of_prepare_op(const exprt &op)
 {
   exprt result = op;
-  if(op.id() == ID_typecast)
+  while(result.id() == ID_typecast)
   {
     result = result.op0();
   }
-  if(
-    result.op0().id() == ID_index &&
-    to_index_expr(result.op0()).index().is_zero())
+  POSTCONDITION(result.id() == ID_address_of);
+  return result;
+}
+
+// returns true if op is a symbol or an expression like x.y[constant].z
+static bool simplify_inequality_is_simple_structure_expression(const exprt &op)
+{
+  exprt expr = op;
+  while(expr.id() == ID_member || expr.id() == ID_index)
   {
-    result = address_of_exprt(to_index_expr(result.op0()).array());
+    if(expr.id() == ID_member)
+    {
+      expr = to_member_expr(expr).compound();
+    }
+    if(expr.id() == ID_index)
+    {
+      const index_exprt index = to_index_expr(expr);
+      if(!index.index().is_constant())
+      {
+        return false;
+      }
+      expr = index.array();
+    }
   }
-  return result;
+  return expr.id() == ID_symbol;
 }
 
 bool simplify_exprt::simplify_inequality_address_of(exprt &expr)
 {
-  assert(expr.type().id()==ID_bool);
-  assert(expr.operands().size()==2);
-  assert(expr.id()==ID_equal || expr.id()==ID_notequal);
+  PRECONDITION(expr.type().id() == ID_bool);
+  PRECONDITION(expr.operands().size() == 2);
+  PRECONDITION(expr.id() == ID_equal || expr.id() == ID_notequal);
 
   exprt lhs = simplify_inequality_address_of_prepare_op(expr.op0());
   exprt rhs = simplify_inequality_address_of_prepare_op(expr.op1());
-  assert(lhs.id() == ID_address_of);
-  assert(rhs.id() == ID_address_of);
 
   if(lhs.operands().size() != 1 || rhs.operands().size() != 1)
     return true;
 
-  if(lhs.op0().id() == ID_symbol && rhs.op0().id() == ID_symbol)
-  {
-    bool equal = lhs.op0().get(ID_identifier) == rhs.op0().get(ID_identifier);
+  // strip the address-of operator
+  lhs = lhs.op0();
+  rhs = rhs.op0();
 
+  if(
+    simplify_inequality_is_simple_structure_expression(lhs) &&
+    simplify_inequality_is_simple_structure_expression(rhs))
+  {
+    object_descriptor_exprt lhs_descriptor, rhs_descriptor;
```

Comment:
With this work, I'm not sure I see a reason for the first commit (or indeed the `simplify_inequality_address_of_prepare_op`)?

Reply:
I'm stripping the typecasts for `is_simple_structure_expression`. It's probably not necessary to have the refactoring as a separate commit, it was only that way because I did that first before making any changes, I can squash all of that into one commit if you prefer.

Reply:
Why is the call to `simplify_inequality_is_simple_structure_expression` even necessary in the first place?

Re squashing: possibly, but there will be room for further changes as the entire preparation before constructing the `object_descriptor_exprt` now seems unnecessary (whether in the original or in the refactored form).

Reply:
Regarding `is_simple_structure_expression`; This is there to check that the expression is something that is safe to simplify (as opposed to constructs like `&random_struct()->x == &random_struct()->x` - in that case, the `root_object`s would be `*random_struct()`, and the offsets would both be constant and equal, so without the `is_simple_structure_expression` with this logic the simplifier would (incorrectly) simplify this to `true`.